ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/DB/Read.hs

```diff
@@ -0,0 +1,78 @@
+module Ariadne.Wallet.Cardano.Kernel.DB.Read (
+    -- * Read-only, pure getters
+    accountUtxo
+  , accountAvailableUtxo
+  , accountAvailableBalance
+  , accountTotalBalance
+  , accountAddresses
+  , hdWallets
+  ) where
+
+import Data.Text.Buildable (Buildable)
+import Formatting (build, sformat)
+
+import Pos.Core (Coin)
+import Pos.Txp (Utxo)
+
+import Ariadne.Wallet.Cardano.Kernel.DB.AcidState (DB, dbHdWallets)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
+  (HdAccountId, HdAddress, HdWallets)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet.Read
+  (HdQueryErr, readAddressesByAccountId)
+import qualified Ariadne.Wallet.Cardano.Kernel.DB.Spec.Read as Spec
+import Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet (IxSet)
+
+{-------------------------------------------------------------------------------
+                              Wallet getters
+
+  The @only@ effectful function we expose is 'getWalletSnapshot', which reads
+  the full DB 'Snapshot' and returns it.
+  All the other getters are completely pure and take the 'Snapshot' as input,
+  so that users of the wallet are forced to re-use the same 'Snapshot' in case
+  they want to read the state of the wallet multiple times within the same
+  code block / handler.
+
+-------------------------------------------------------------------------------}
+
+walletQuery' :: forall e a. (Buildable e)
+             => DB
+             -> HdQueryErr e a
+             -> a
+walletQuery' snapshot qry= do
+    let res = qry (snapshot ^. dbHdWallets)
+    either err identity res
+    where
+        err = error . sformat build
```

Comment:
I don't like this whole module because of this function which calls `error` when `qry` fails.
`error` should be used only for programmer mistakes.
Functions from this module shouldn't assume that each fail of `qry` is a programmer mistake.
The only function from here that you use is `accountAvailableUtxo`. I propose to move to the place where it's used and get rid of this module.
The more localized `error` is, the better.